File: pages/coingecko.js
import { By } from 'selenium-webdriver';
import { sendErrorToTelegram } from '../utils/errorToTelegram.js';
import process from 'process';

const handlerError = async (e, driver, addErrorMsg) => {
  console.error(addErrorMsg + e);
  await driver.get(driver.getCurrentUrl());
  await sendErrorToTelegram(e, addErrorMsg, process.env.MONITORING_ID_BOT_WASB);
}

const getValueTab = async (driver) => await driver.findElements(By.css('tbody')); 

export const getMarketCap = async (metrics, driver, maxLoop, crypto) => {
  try {
    while(!metrics.marketCap) {
      const valueTab = await getValueTab(driver);
      const marketCap = await valueTab[1].findElements(By.css('td'));
      metrics.marketCap = await marketCap[0].getText()
      if(metrics.marketCap) metrics.marketCap = metrics.marketCap.replaceAll(',', '');

      if(maxLoop === 0) throw new Error(`Nb loop max in getMarketCap ${crypto}.`); 
      maxLoop--;
    }

    maxLoop = 5;
    console.log(crypto + ' marketCap is acquired.');
  } catch(e) {
    await handlerError(e, driver, `Error to get ${crypto} marketCap on CoinGecko: `);
  }
}

// Get 24h volume.
export const getVolume = async (metrics, driver, maxLoop, crypto) => {
  try {
    while(!metrics.volumeCoinGecko) {
      const valueTab = await getValueTab(driver);
      const volume = await valueTab[1].findElements(By.css('td'));
      metrics.volumeCoinGecko = await volume[3].getText();

      if(maxLoop === 0) throw new Error(`Nb loop max in getVolume ${crypto}.`); 
      maxLoop--;
    }

    maxLoop = 5;
    console.log(crypto + ' volume is acquired.');
  } catch(e) {
    await handlerError(e, driver, `Error to get ${crypto} volume on CoinGecko: `);
  }
}

// Get supply circulation.
export const getSupplyCirculation = async (metrics, driver, maxLoop, crypto) => {
  try {
    while(!metrics.supplyCirculation) {
      const valueTab = await getValueTab(driver);
      const supplyCirculation = await valueTab[1].findElements(By.css('td'));
      metrics.supplyCirculation = await supplyCirculation[4].getText();
      if(metrics.supplyCirculation) metrics.supplyCirculation = metrics.supplyCirculation.replaceAll(',', '');

      if(maxLoop === 0) throw new Error(`Nb loop max in getSupplyCirculation ${crypto}.`); 
      maxLoop--;
    }

    maxLoop = 5;
    console.log(crypto + ' supply circulation is acquired.');
  } catch(e) {
    await handlerError(e, driver, `Error to get ${crypto} supply circulation on CoinGecko: `);
  }
}

// Get max supply.
export const getMaxSupply = async (metrics, driver, maxLoop, crypto) => {
  try {
    while(!metrics.maxSupply) {
      const valueTab = await getValueTab(driver);
      const maxSupply = await valueTab[1].findElements(By.css('td'));
      metrics.maxSupply = await maxSupply[6].getText();

      if(maxLoop === 0) throw new Error(`Nb loop max in getMaxSupply ${crypto}.`); 
      maxLoop--;
    }

    maxLoop = 5;
    console.log(crypto + ' maxSupply is acquired.');
  } catch(e) {
    await handlerError(e, driver, `Error to get ${crypto} maxSupply on CoinGecko: `);
  }
}

// Get rank crypto.
export const getRank = async (crypto, driver, maxLoop) => {
  try {
    while(!crypto.rank) {
      const rank = await driver.findElement(By.css('[data-coin-show-target="staticCoinPrice"] div > span > div'));
      crypto.rank = await rank.getText();
      crypto.rank = crypto.rank.split('#')[1];

      if(maxLoop === 0) throw new Error('Nb loop max in getRank.');
      maxLoop--;
    }

    maxLoop = 5;
    console.log('Rank is acquired.');
  } catch(e) {
    await handlerError(e, driver, 'Error for get rank: ');
  }
}

File: pages/coingecko.py
from playwright.async_api import Page
from utils.telegram import handler_error

def get_value_tab(page: Page): return page.locator('tbody').locator('nth=1').locator('td')

async def get_market_cap(metrics: dict, page: Page, max_loop: int, crypto: str):
  try:
    while metrics['marketCap'] == '':
      market_cap = get_value_tab(page)
      metrics['marketCap'] = await market_cap.first.text_content()
      if metrics['marketCap']: metrics['marketCap'] = metrics['marketCap'].replace(',', '').strip()

      if max_loop == 0: raise Exception(f'Nb loop max in get_market_cap {crypto}.')
      max_loop = max_loop -1

    max_loop = 5
    print(crypto + ' marketCap is acquired.')
  except Exception as e:
    await handler_error(e, page, f'Error to get {crypto} marketCap on CoinGecko: ')

# Get 24h volume.
async def get_volume(metrics: dict, page: Page, max_loop: int, crypto: str):
  try:
    while metrics['volumeCoinGecko'] == '':
      volume = get_value_tab(page)
      metrics['volumeCoinGecko'] = await volume.locator('nth=3').text_content()
      metrics['volumeCoinGecko'] = metrics['volumeCoinGecko'].strip()

      if max_loop == 0: raise Exception(f'Nb loop max in get_volume {crypto}.')
      max_loop = max_loop -1

    max_loop = 5
    print(crypto + ' volume is acquired.')
  except Exception as e:
    await handler_error(e, page, f'Error to get {crypto} volume on CoinGecko: ')

# Get supply circulation.
async def get_supply_circulation(metrics: dict, page: Page, max_loop: int, crypto: str):
  try:
    while metrics['supplyCirculation'] == '':
      supply_circulation = get_value_tab(page)
      metrics['supplyCirculation'] = await supply_circulation.locator('nth=4').text_content()
      if metrics['supplyCirculation']: 
        metrics['supplyCirculation'] = metrics['supplyCirculation'].replace(',', '').strip()
        if '\n' in metrics['supplyCirculation']: metrics['supplyCirculation'] =  metrics['supplyCirculation'].split('\n')[0]

      if max_loop == 0: raise Exception(f'Nb loop max in get_supply_circulation {crypto}.')
      max_loop = max_loop -1

    max_loop = 5
    print(crypto + ' supply circulation is acquired.')
  except Exception as e:
    await handler_error(e, page, f'Error to get {crypto} supply circulation on CoinGecko: ')

# Get rank crypto.
async def get_rank(crypto: str, page: Page, max_loop: int):
  try:
    while crypto['rank'] == '':
      rank = page.locator('[data-coin-show-target="staticCoinPrice"]').locator('div').locator('span').locator('div')
      crypto['rank'] = await rank.text_content()
      crypto['rank'] = crypto['rank'].split('#')[1].strip()

      if max_loop == 0: raise Exception('Nb loop max in get_rank.')
      max_loop = max_loop -1

    max_loop = 5
    print('Rank is acquired.')
  except Exception as e:
    await handler_error(e, page, 'Error for get rank: ')